File: eco_restoration_shard/cyboquatic_progress/20260714/cpp/drainagedecay_bod_tss_cec.cpp
// File: eco_restoration_shard/cyboquatic_progress/20260714/cpp/drainagedecay_bod_tss_cec.cpp
// Domain (e): drainagedecay frames (BOD, TSS, CEC) for cyboquatic machinery.
// Standard C++17 code; no external dependencies. Build with: g++ -std=c++17 -O2 -o drainagedecay_bod_tss_cec drainagedecay_bod_tss_cec.cpp

#include <cmath>
#include <cfloat>
#include <iostream>
#include <stdexcept>
#include <string>

struct DrainageState {
    double bodMgL;
    double tssMgL;
    double cecCmolKg;
    double temperatureC;
    double flowLps;
};

struct DecayParameters {
    double kBodPerDay;
    double kTssPerDay;
    double theta;
    double refTempC;
};

static double firstOrderDecay(double initial, double kPerHour, double dtHours) {
    if (initial <= 0.0) return 0.0;
    if (kPerHour <= 0.0 || dtHours == 0.0) return initial;
    double exponent = -kPerHour * dtHours;
    return initial * std::exp(exponent);
}

static double temperatureFactor(double theta, double refTempC, double currentTempC) {
    double delta = currentTempC - refTempC;
    return std::exp(std::log(theta) * (delta / 10.0));
}

static DrainageState step(const DrainageState &state, const DecayParameters &params, double dtHours) {
    if (dtHours < 0.0) {
        throw std::invalid_argument("dtHours must be non-negative");
    }

    double tempFactor = temperatureFactor(params.theta, params.refTempC, state.temperatureC);

    double kBodPerHour = params.kBodPerDay / 24.0 * tempFactor;
    double kTssPerHour = params.kTssPerDay / 24.0 * tempFactor;

    double bodNext = firstOrderDecay(state.bodMgL, kBodPerHour, dtHours);
    double tssNext = firstOrderDecay(state.tssMgL, kTssPerHour, dtHours);

    DrainageState result = state;
    result.bodMgL = bodNext < 0.0 ? 0.0 : bodNext;
    result.tssMgL = tssNext < 0.0 ? 0.0 : tssNext;

    return result;
}

static double oxygenDemandMgPerSec(const DrainageState &state) {
    double bodNonNegative = state.bodMgL < 0.0 ? 0.0 : state.bodMgL;
    double flowNonNegative = state.flowLps < 0.0 ? 0.0 : state.flowLps;
    return bodNonNegative * flowNonNegative / 1000.0;
}

static std::string toString(const DrainageState &s) {
    std::string out;
    out += "DrainageState{";
    out += "bodMgL=" + std::to_string(s.bodMgL);
    out += ", tssMgL=" + std::to_string(s.tssMgL);
    out += ", cecCmolKg=" + std::to_string(s.cecCmolKg);
    out += ", temperatureC=" + std::to_string(s.temperatureC);
    out += ", flowLps=" + std::to_string(s.flowLps);
    out += "}";
    return out;
}

int main() {
    DrainageState state{45.0, 90.0, 30.0, 23.0, 6.0};
    DecayParameters params{0.22, 0.08, 1.05, 20.0};
    double dtHours = 3.0;

    DrainageState next = step(state, params, dtHours);
    double oxygenDemand = oxygenDemandMgPerSec(next);

    std::cout << "Initial: " << toString(state) << "\n";
    std::cout << "Next after " << dtHours << " h: " << toString(next) << "\n";
    std::cout << "Oxygen demand (mg O2/s): " << oxygenDemand << "\n";

    return 0;
}

File: eco_restoration_shard/cyboquatic_progress/20260714/java/DrainagedecayBodTssCec.java
// File: eco_restoration_shard/cyboquatic_progress/20260714/java/DrainagedecayBodTssCec.java
// Domain (e): drainagedecay frames (BOD, TSS, CEC) for cyboquatic machinery.
// Plain Java 11+ compatible, with no external dependencies.

package cyboquatic.drainagedecay;

public final class DrainagedecayBodTssCec {

    public static final class DrainageState {
        public final double bodMgL;
        public final double tssMgL;
        public final double cecCmolKg;
        public final double temperatureC;
        public final double flowLps;

        public DrainageState(
                double bodMgL,
                double tssMgL,
                double cecCmolKg,
                double temperatureC,
                double flowLps
        ) {
            this.bodMgL = bodMgL;
            this.tssMgL = tssMgL;
            this.cecCmolKg = cecCmolKg;
            this.temperatureC = temperatureC;
            this.flowLps = flowLps;
        }

        public DrainageState copy(
                Double bodMgLOverride,
                Double tssMgLOverride
        ) {
            double bodValue = bodMgLOverride != null ? bodMgLOverride : this.bodMgL;
            double tssValue = tssMgLOverride != null ? tssMgLOverride : this.tssMgL;
            return new DrainageState(
                    bodValue,
                    tssValue,
                    this.cecCmolKg,
                    this.temperatureC,
                    this.flowLps
            );
        }

        @Override
        public String toString() {
            return "DrainageState{" +
                    "bodMgL=" + bodMgL +
                    ", tssMgL=" + tssMgL +
                    ", cecCmolKg=" + cecCmolKg +
                    ", temperatureC=" + temperatureC +
                    ", flowLps=" + flowLps +
                    '}';
        }
    }

    public static final class DecayParameters {
        public final double kBodPerDay;
        public final double kTssPerDay;
        public final double theta;
        public final double refTempC;

        public DecayParameters(
                double kBodPerDay,
                double kTssPerDay,
                double theta,
                double refTempC
        ) {
            this.kBodPerDay = kBodPerDay;
            this.kTssPerDay = kTssPerDay;
            this.theta = theta;
            this.refTempC = refTempC;
        }
    }

    public static DrainageState step(
            DrainageState state,
            DecayParameters params,
            double dtHours
    ) {
        if (dtHours < 0.0) {
            throw new IllegalArgumentException("dtHours must be non-negative");
        }

        double tempFactor = temperatureFactor(
                params.theta,
                params.refTempC,
                state.temperatureC
        );

        double kBodPerHour = params.kBodPerDay / 24.0 * tempFactor;
        double kTssPerHour = params.kTssPerDay / 24.0 * tempFactor;

        double bodNext = firstOrderDecay(state.bodMgL, kBodPerHour, dtHours);
        double tssNext = firstOrderDecay(state.tssMgL, kTssPerHour, dtHours);

        double bodClamped = Math.max(0.0, bodNext);
        double tssClamped = Math.max(0.0, tssNext);

        return state.copy(bodClamped, tssClamped);
    }

    public static double oxygenDemandMgPerSec(DrainageState state) {
        double bodNonNegative = Math.max(0.0, state.bodMgL);
        double flowNonNegative = Math.max(0.0, state.flowLps);
        return bodNonNegative * flowNonNegative / 1000.0;
    }

    private static double firstOrderDecay(
            double initial,
            double kPerHour,
            double dtHours
    ) {
        if (initial <= 0.0) return 0.0;
        if (kPerHour <= 0.0 || dtHours == 0.0) return initial;
        double exponent = -kPerHour * dtHours;
        return initial * Math.exp(exponent);
    }

    private static double temperatureFactor(
            double theta,
            double refTempC,
            double currentTempC
    ) {
        double delta = currentTempC - refTempC;
        return Math.exp(Math.log(theta) * (delta / 10.0));
    }

    public static void main(String[] args) {
        DrainageState initial = new DrainageState(
                35.0,
                60.0,
                20.0,
                21.0,
                4.0
        );

        DecayParameters params = new DecayParameters(
                0.18,
                0.07,
                1.05,
                20.0
        );

        double dtHours = 4.0;

        DrainageState next = step(initial, params, dtHours);
        double oxygenDemand = oxygenDemandMgPerSec(next);

        System.out.println("Initial: " + initial);
        System.out.println("Next after " + dtHours + " h: " + next);
        System.out.println("Oxygen demand (mg O2/s): " + oxygenDemand);
    }
}
